## Homework - Machine Learning Preparation
### Agi Rahmawandi <br> Batch 57



In [2]:
import warnings
warnings.filterwarnings('ignore')

import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
%matplotlib inline


print('numpy version : ',np.__version__)
print('pandas version : ',pd.__version__)
print('seaborn version : ',sns.__version__)

numpy version :  2.2.5
pandas version :  2.2.3
seaborn version :  0.13.2


In [17]:
sns.set(rc={'figure.figsize':(20.7,8.27)})
sns.set_style("whitegrid")
sns.color_palette("dark")
plt.style.use("fivethirtyeight")


**Product Classification**<br>
**Deskripsi**<br>
* Memprediksi apakah suatu produk eksklusif atau tidak berdasarkan fitur yang tersedia<br>


### Load data

In [4]:
df = pd.read_csv('./dataset/Product_Exclusive_Classification.csv')

## 1. Descriptive Statistics (5 poin)
Gunakan function info dan describe pada dataset final project kalian. Tuliskan hasil
observasinya, seperti:<br>
A. Apakah ada kolom dengan tipe data kurang sesuai, atau nama kolom dan isinya
kurang sesuai?<br>
B. Apakah ada kolom yang memiliki nilai kosong? Jika ada, apa saja?<br>
C. Apakah ada kolom yang memiliki nilai summary agak aneh?<br>
(min/mean/median/max/unique/top/freq)<br>
* Untuk masing-masing jenis observasi, tuliskan juga

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8000 entries, 0 to 7999
Data columns (total 9 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   id                 8000 non-null   int64  
 1   brand              8000 non-null   object 
 2   category           7987 non-null   object 
 3   rating             7905 non-null   float64
 4   number_of_reviews  7991 non-null   float64
 5   love               7966 non-null   float64
 6   price              7992 non-null   float64
 7   value_price        7983 non-null   float64
 8   exclusive          8000 non-null   int64  
dtypes: float64(5), int64(2), object(2)
memory usage: 562.6+ KB


In [11]:
## pengelompokan berdasar jenis kolom
obj = ['brand','category']
numbs = ['id','rating','number_of_reviews','love','price','value_price']

In [13]:
df[obj].describe()

brand category
count                 8000     7987
unique                 310      142
top     SEPHORA COLLECTION  Perfume
freq                   492      619

In [14]:
df[numbs].describe()

id       rating  number_of_reviews          love  \
count  8.000000e+03  7905.000000        7991.000000  7.966000e+03   
mean   1.910231e+06     4.085136         303.574396  1.756396e+04   
std    3.858353e+05     0.761069         931.724460  4.425339e+04   
min    5.000000e+01     0.000000           0.000000  0.000000e+00   
25%    1.773379e+06     4.000000          14.000000  2.000000e+03   
50%    2.030360e+06     4.000000          56.000000  5.500000e+03   
75%    2.185074e+06     4.500000         231.500000  1.530000e+04   
max    2.293801e+06     5.000000       19000.000000  1.300000e+06   

             price  value_price  
count  7992.000000  7983.000000  
mean     49.900935    50.983300  
std      46.864764    48.473049  
min       2.000000     2.000000  
25%      24.000000    24.000000  
50%      35.000000    35.000000  
75%      59.000000    60.000000  
max     549.000000   549.000000

In [8]:
df.isna().sum()

id                    0
brand                 0
category             13
rating               95
number_of_reviews     9
love                 34
price                 8
value_price          17
exclusive             0
dtype: int64

In [15]:
df.sample(5)

id               brand      category  rating  number_of_reviews  \
4644  2077485  Kiehl's Since 1851      Skincare     3.0                5.0   
7045  2242105         MILK MAKEUP    Foundation     4.0              924.0   
4738  2082766  SEPHORA COLLECTION  Moisturizers     3.5               19.0   
5372  2132546   Charlotte Tilbury   Highlighter     4.0              123.0   
6778  2224053        IT Cosmetics     Mini Size     3.5               46.0   

         love  price  value_price  exclusive  
4644   2600.0   13.0         13.0          0  
7045  26200.0   36.0         36.0          1  
4738   1800.0   17.0         17.0          1  
5372  31400.0   58.0         58.0          0  
6778   6400.0   12.0         12.0          0

A. Type data dan nama kolom rata-rata sudah sesuai, hanya saja akan lebih baik untuk id di ubah ke obj<br>
B. kolom yang memiliki nilai kosong adalah category,rating, number_of_reviews,love,price,value_price<br>
C. dikolom number_of_reviews nilai mean dan median terlihat cukup jauh perbedaan nya **mean 303.57**, **median 56** serta **standar deviasinya** juga rentangnya cukup jauh dari mean yakni **931.7**

https://www.kaggle.com/code/mohammadfikri/product-exclusive-classification-roc-auc-87/notebook

# Exploratory Data Analysis (EDA)

## Pendekatan Numerik (10 Poin)

## Pendekatan Grafis

## 2. Univariate Analysis
Gunakan visualisasi untuk melihat distribusi masing-masing kolom (feature maupun
target). <br>Tuliskan hasil observasinya, misalnya jika ada suatu kolom yang distribusinya
menarik (misal skewed, bimodal, ada outlier, ada nilai yang mendominasi, kategorinya
terlalu banyak, dsb).<br> Jelaskan juga apa yang harus di-follow up saat data pre-processing.

In [24]:
for col in range in obj:
    print(f'''Value count kolom {col}:''')
    print(df[col].value_counts())
    prit()

TypeError: 'bool' object is not iterable

In [31]:
for i in range(0,len(numbs)):
    plt.subplot(1,len(numbs),i+1)
    sns.boxplot(y=df[numbs[i]],color='gray',orient='v')
    plt.tight_layout()

In [32]:
for i in range(0, len(numbs)):
    plt.subplot(1, len(numbs), i+1)
    sns.violinplot(y=df[numbs[i]], color='gray', orient='v')
    plt.tight_layout()

## 3. Multivariate Analysis (15 poin)
Lakukan multivariate analysis (seperti correlation heatmap dan category plots, sesuai yang
diajarkan di kelas). Tuliskan hasil observasinya, seperti: <br>
A. Bagaimana korelasi antara masing-masing feature dan label. Kira-kira feature mana
saja yang paling relevan dan harus dipertahankan?<br>
B. Bagaimana korelasi antar-feature, apakah ada pola yang menarik? Apa yang perlu
dilakukan terhadap feature itu?<br>
* Tuliskan juga jika memang tidak ada feature yang saling berkorelasi

## Insight (20 Poin)

Apa saja insight yang kamu temukan dari hasil-hasil analisis di atas?

## 4. Data Cleansing (40 poin)
Lakukan pembersihan data, sesuai yang diajarkan di kelas, seperti:<br>
A. Handle missing values<br>
B. Handle duplicate data<br>
C. Handle outliers<br>
D. Feature transformation<br>
E. Feature encoding<br>
F. Handle class imbalance<br>
Di laporan homework, tuliskan apa saja yang telah dilakukan dan metode yang digunakan.<br>
* Tetap tuliskan jika memang ada tidak yang perlu di-handle (contoh: “Tidak perlu feature
encoding karena semua feature sudah numerical” atau “Outlier tidak di-handle karena akan
fokus menggunakan model yang robust terhadap outlier”).

### A. Missing Value 

### B. Duplicate Values 


### C. Outliers 

### D.Feature transformation


### E. Feature encoding

### F. Handle class imbalance

# 5. Feature Engineering (30 poin)
Cek feature yang ada sekarang, lalu lakukan:<br>
A. Feature selection (membuang feature yang kurang relevan atau redundan)<br>
B. Feature extraction (membuat feature baru dari feature yang sudah ada)<br>
C. Tuliskan minimal 4 feature tambahan (selain yang sudah tersedia di dataset) yang
mungkin akan sangat membantu membuat performansi model semakin bagus (ini hanya
ide saja, untuk menguji kreativitas teman-teman, tidak perlu benar-benar dicari datanya
dan tidak perlu diimplementasikan)<br>
* Untuk 5A & 5B, tetap tuliskan jika memang tidak bisa dilakukan (contoh: “Semua feature
digunakan untuk modelling (tidak ada yang dihapus), karena semua feature relevan”)<br>

### A. Feature selection

### B. Feature extraction

### C. feature tambahan